### PR TITLE
* fix config.url startswith slash

### PR DESCRIPTION
**Pull request by UMU618**
If I use '//blog.umu618.com' as config.url, the url_for is malfunctioned.

### CODE REVIEW

File: lib/full_url_for.js [hunk 1]
```diff
@@ -18,7 +18,8 @@ function fullUrlForHelper(path = '/') {
   return cache.apply(`${config.url}-${prettyUrlsOptions.trailing_index}-${prettyUrlsOptions.trailing_html}-${path}`, () => {
     if (/^(\/\/|http(s)?:)/.test(path)) return path;
 
-    const sitehost = parse(config.url).hostname || config.url;
+    const config_url = config.url.startsWith('//') ? `https:${config.url}` : config.url;
```
Comment (SukkaW) [inline lib/full_url_for.js]: Why `//` prefixed URL will be considered as `https` supported?

**Comment by UMU618** [comment]
> I can't see the point of the PR.
> 
> If you are going to address a fix, then add your test cases then.

if you use HTTP and HTTPS together, you may set config.url as //your-domain.

**Comment by curbengh** [comment]
> if you use HTTP and HTTPS together

May I know your specific use cases of `config.url`? For default use case, it's mainly for SEO purpose ([open_graph.js](https://github.com/hexojs/hexo/blob/master/lib/plugins/helper/open_graph.js)). SEO typically requires a canonical link, either HTTP or HTTPS, not both.

For linking other pages in your blog, you should use relative link, e.g. `/2018/03/20/foo-bar/`, homepage is just `/`.

**Comment by UMU618** [comment]
> > if you use HTTP and HTTPS together
> 
> May I know your specific use cases of `config.url`? For default use case, it's mainly for SEO purpose ([open_graph.js](https://github.com/hexojs/hexo/blob/master/lib/plugins/helper/open_graph.js)). SEO typically requires a canonical link, either HTTP or HTTPS, not both.
> 
> For linking other pages in your blog, you should use relative link, e.g. `/2018/03/20/foo-bar/`, homepage is just `/`.

## * fix config.url startswith slash #188
## https://github.com/hexojs/hexo-util/pull/188
url: //blog.umu618.com

**Comment by UMU618** [comment]
> > if you use HTTP and HTTPS together
> 
> May I know your specific use cases of `config.url`? For default use case, it's mainly for SEO purpose ([open_graph.js](https://github.com/hexojs/hexo/blob/master/lib/plugins/helper/open_graph.js)). SEO typically requires a canonical link, either HTTP or HTTPS, not both.
> 
> For linking other pages in your blog, you should use relative link, e.g. `/2018/03/20/foo-bar/`, homepage is just `/`.

https://github.com/UMU618/blog/blob/master/_config.yml

**Review by stevenjoezhang — CHANGES_REQUESTED** [review]
It is recommended to redirect http to https with status code 301

**Comment by UMU618** [comment]
> It is recommended to redirect http to https with status code 301

Agree. but not all free hosts support https.

**Comment by UMU618** [comment]
> If https is not supported, the URL should be `http://your-domain`, instead of `//your-domain` in Hexo `_config.yml`

Maintaining two copies of config is inconvenient, so I modify the code.

**Comment by UMU618** [comment]
> I don't understand why it's necessary maintaining two copies of config.
> 
> * If your site supports HTTPS: `url: https://your-domain`
> * If your site does not support HTTPS: `url: http://your-domain`
> * If your site supports both HTTP and HTTPS: `url: https://your-domain` and set 301 redirect
> 
> In all these cases, one config file is enough.

What about having 2 hosts, one supports HTTPS, another not?

**Comment by stevenjoezhang** [comment]
If you have two hosts serving the same content, I would suggest a 301 redirect for better SEO.

**Comment by stevenjoezhang** [comment]
Besides, even if you have successfully set '//blog.umu618.com' as config.url, you still can not deploy it to another host. Your Pull Request does not resolve the issue.

**Comment by UMU618** [comment]
> If you have two hosts serving the same content, I would suggest a 301 redirect for better SEO.

Thank you. People that can access host A, can NOT access host B.

**Comment by UMU618** [comment]
> Besides, even if you have successfully set '//blog.umu618.com' as config.url, you still can not deploy it to another host. Your Pull Request does not resolve the issue.

deploy to another host via `scp`

**Comment by jiangtj** [comment]
> People that can access host A, can NOT access host B.

Maybe consider overriding configuration parameters via the command line. For example, to implement the following command line.

```
hexo g --config.url=http(s)://another.host
```

Currently, Hexo supports custom configuration file paths

image

**Comment by UMU618** [comment]
这个 PR 很显然是有用的，您在本地测试的时候一般就用 http，上传到服务器上就用 https，这就是典型的两处环境不一样，都兼容不是更好的？

**Comment by stevenjoezhang** [comment]
你所说的这个情况本来就支持。设置 `url: https://your-domain` 就可以实现你说的

> 在本地测试的时候一般就用 http，上传到服务器上就用 https

本地测试打开 `http://localhost:4000`，上传服务器就用 `https://your-domain`，没有任何问题。

**Comment by UMU618** [comment]
> 你所说的这个情况本来就支持。设置 `url: https://your-domain` 就可以实现你说的

我的本地使用的一毛一样的域名，设置 hosts 的，不是 localhost，因为我有两个网站，一个不支持 https。

**Comment by stevenjoezhang** [comment]
所以你的根本需求还是：有两个域名，一个支持 https，一个不支持，它们需要部署同一个 Hexo 博客。是这样么？
不然本地测试并不需要设置 hosts，这样反而麻烦了。

**Comment by UMU618** [comment]
> 所以你的根本需求还是：有两个域名，一个支持 https，一个不支持，它们需要部署同一个 Hexo 博客。是这样么？
> 不然本地测试并不需要设置 hosts，这样反而麻烦了。

是这样的，前面用英文讲过，被忽视，所以只能大家能遇到的举个例子。

**Comment by stevenjoezhang** [comment]
我觉得没有被忽视，我已经给出过建议

截屏2022-02-15 下午10 34 18

所以这个不支持 https 的域名是你只在本地测试时使用，还是可以通过公网访问？

**Comment by UMU618** [comment]
您可以理解为 CDN 的贫穷版……自然是要公网访问，域名可以对不同运营商解析不同 IP，这也可能导致页面放在多个 host 上。

**Pull request closed without merging**